gce/util.sh

```diff
@@ -9,6 +9,171 @@
+	util::exec_with_retry "gcloud compute instance-templates create ${GK_TEMPLATE} \
+		--image=$CLUSTER_OS_IMAGE --image-project=$CLUSTER_OS_IMAGE_PROJECT \
+		--machine-type=$MACHINE_TYPE --network=$GCP_NETWORK \
+		--subnet=$GCP_NETWORK --region=$GCP_REGION  \
+		--boot-disk-auto-delete --boot-disk-size=$NODE_BOOT_DISK_SIZE \
+		--boot-disk-type=$NODE_BOOT_DISK_TYPE \
+		--metadata-from-file=\"startup-script\"=$STARTUP_SCRIPT" $RETRY_MAX
+	return 0
+}
+
+function __delete_instance_template() {
+	GK_TEMPLATE="$GK_NODE_NAME"
+	echo "-- Looking for old templates."
+	if gcloud compute instance-templates describe $GK_TEMPLATE &>/dev/null; then
+		echo "-- Instance template $GK_TEMPLATE already exists. Checking for dependent instance groups."
+		# Cleanup old groups.  Templates cannot be deleted until they have no dependent groups..
+		if gcloud compute instance-groups managed describe $GK_NODE_NAME \
+			--zone=$GCP_ZONE &>/dev/null; then
+			echo "-- Instance group $GK_NODE_NAME already exists, deleting before proceeding."
+			if ! gcloud compute instance-groups managed delete $GK_NODE_NAME \
+				--zone=$GCP_ZONE --quiet; then
+				echo "-- Failed to delete instance group $GK_NODE_NAME."
+				return 1
+			fi
+		fi
+		echo "-- Deleting instance template $GK_TEMPLATE."
+		if ! gcloud compute instance-templates delete $GK_TEMPLATE --quiet; then
+			echo "-- Failed to delete instance-template $GK_TEMPLATE."
+			return 1
+		fi
+	else
+		echo "-- No pre-existing template found."
+	fi
+	return 0
+}
+
+function __create_instance_group() {
+	echo "-- Creating instance group: $GK_NODE_NAME."
+	util::exec_with_retry "gcloud compute instance-groups managed create $GK_NODE_NAME --zone=$GCP_ZONE \
+		--template=$GK_TEMPLATE --size=$GK_NUM_NODES" $RETRY_MAX
+	return 0
+}
+
+function __delete_instance_group() {}
+
+function __create_master() {
+	echo "-- Looking for old master instance: $GK_MASTER_NAME."
+	if gcloud compute instances describe $GK_MASTER_NAME --zone=$GCP_ZONE &>/dev/null; then
+		echo "-- Instance $GK_MASTER_NAME  already exists. Deleting it before proceeding."
+		if ! gcloud compute instances delete $GK_MASTER_NAME --zone=$GCP_ZONE --quiet; then
+			echo "-- Failed to delete instance $GK_MASTER_NAME"
+			return 1
+		fi
+	else
+		echo "-- No pre-existing master instance found."
+	fi
+
+	echo "-- Creating master instance: $GK_MASTER_NAME"
+	util::exec_with_retry "gcloud compute instances create $GK_MASTER_NAME --boot-disk-auto-delete \
+		--boot-disk-size=$NODE_BOOT_DISK_SIZE --boot-disk-type=$NODE_BOOT_DISK_TYPE \
+		--image-project=$CLUSTER_OS_IMAGE_PROJECT --machine-type=$MASTER_MACHINE_TYPE \
+		--network=$GCP_NETWORK --zone=$GCP_ZONE --image=$CLUSTER_OS_IMAGE \
+		--metadata-from-file=\"startup-script\"=$STARTUP_SCRIPT" $RETRY_MAX \
+	|| return 1
+	return 0
+}
+
+
+function __create_secondary_disks() {
+	echo "-- Creating RHGS block devices: ${OBJ_STORAGE_ARR[@]}"
+	util::exec_with_retry "gcloud compute disks create ${OBJ_STORAGE_ARR[*]} \
```

Comment:
If this util::exec_with_retry lives in _lib/_ then we have to figure out if we want it there (and then change the name prefix to "lib::") or if thity should be in _gce/util.sh_"

Reply:
Uses global vars which may be necessary but, IMO, should be avoided if reasonable.

Reply:
Right.  We shouldn't put it in either provider util though, so maybe a common util that both providers import.

Reply:
Yes, I was also thinking we probably need a new _common.sh_ in repo root (or I guess it could live in _lib/_)